pyhf/pdf.py

```diff
@@ -70,68 +88,146 @@ def par_slice(self, name):
-        """
-        Add a new modifier if it does not exist and return it
-        or get the existing modifier and return it
 
-        Args:
-            channel: current channel object (e.g. from spec)
-            sample: current sample object (e.g. from spec)
-            modifier_def: current modifier definitions (e.g. from spec)
-
-        Returns:
-            modifier object
-
-        """
+    def bookkeep_paramsets(self, channel, sample, name, modifier_type):
         # get modifier class associated with modifier type
         try:
-            modifier_cls = modifiers.registry[modifier_def['type']]
+            modifier_cls = modifiers.registry[modifier_type]
         except KeyError:
-            log.exception('Modifier type not implemented yet (processing {0:s}). Current modifier types: {1}'.format(modifier_def['type'], modifiers.registry.keys()))
+            log.exception('Modifier type not implemented yet (processing {0:s}). Current modifier types: {1}'.format(modifier_type, modifiers.registry.keys()))
             raise exceptions.InvalidModifier()
 
-        # if modifier is shared, check if it already exists and use it
-        if modifier_cls.is_shared and modifier_def['name'] in self.par_map:
-            log.info('using existing shared, {0:s}constrained modifier (name={1:s}, type={2:s})'.format('' if modifier_cls.is_constrained else 'un', modifier_def['name'], modifier_cls.__name__))
-            modifier = self.par_map[modifier_def['name']]['modifier']
-            if not type(modifier).__name__ == modifier_def['type']:
-                raise exceptions.InvalidNameReuse('existing modifier is found, but it is of wrong type {} (instead of {}). Use unique modifier names or use qualify_names=True when constructing the pdf.'.format(type(modifier).__name__, modifier_def['type']))
-            return modifier
-
-        # did not return, so create new modifier and return it
-        modifier = modifier_cls(sample['data'], modifier_def['data'])
-        self.register_paramset(modifier_def['name'], modifier.n_parameters, modifier)
-
-        return modifier
+        self.paramsets.setdefault(name, [])
+
+        parset = modifier_cls.required_parset(len(sample['data']))
+        parset.update({'channel,sample': (channel['name'], sample['name'])})
+
+        if not(parset['is_shared']) and self.paramsets.get(name):
+            raise ValueError("Trying to add unshared-paramset but other paramsets exist with the same name.")
+
+        if parset['is_shared'] and self.paramsets.get(name) and not(self.paramsets.get(name)[0]['is_shared']):
+            raise ValueError("Trying to add shared-paramset but other paramset of same name is indicated to be unshared.")
+
+        self.paramsets.get(name).append(parset)
+
+    def combine_paramsets(self):
+        for param_name in list(self.paramsets.keys()):
+            params = self.paramsets[param_name]
+
+            def get_inits(parset, npars, param_matching):
```

Comment:
I would make the inits/bounds/auxdata be part of the requirements returned by the modifier classes.. and here we only check compatibility (e.g. all modifeirs requesting a parmset of a given name need to declare the same bounds, etc)